File: sdks/php/src/NovabillingClient.php
<?php

namespace NovaBilling;

use NovaBilling\Auth\AuthClient;
use NovaBilling\Tenants\TenantsClient;
use NovaBilling\ApiKeys\ApiKeysClient;
use NovaBilling\Currencies\CurrenciesClient;
use NovaBilling\Customers\CustomersClient;
use NovaBilling\Plans\PlansClient;
use NovaBilling\Subscriptions\SubscriptionsClient;
use NovaBilling\Invoices\InvoicesClient;
use NovaBilling\Payments\PaymentsClient;
use NovaBilling\PaymentProviders\PaymentProvidersClient;
use NovaBilling\Webhooks\WebhooksClient;
use NovaBilling\Analytics\AnalyticsClient;
use NovaBilling\Coupons\CouponsClient;
use NovaBilling\AddOns\AddOnsClient;
use NovaBilling\CreditNotes\CreditNotesClient;
use NovaBilling\Portal\PortalClient;
use NovaBilling\BillableMetrics\BillableMetricsClient;
use NovaBilling\Events\EventsClient;
use NovaBilling\Charges\ChargesClient;
use NovaBilling\Wallets\WalletsClient;
use NovaBilling\PaymentMethods\PaymentMethodsClient;
use NovaBilling\Taxes\TaxesClient;
use NovaBilling\PlanOverrides\PlanOverridesClient;
use Psr\Http\Client\ClientInterface;
use NovaBilling\Core\Client\RawClient;

class NovabillingClient
{
    /**
     * @var AuthClient $auth
     */
    public AuthClient $auth;

    /**
     * @var TenantsClient $tenants
     */
    public TenantsClient $tenants;

    /**
     * @var ApiKeysClient $apiKeys
     */
    public ApiKeysClient $apiKeys;

    /**
     * @var CurrenciesClient $currencies
     */
    public CurrenciesClient $currencies;

    /**
     * @var CustomersClient $customers
     */
    public CustomersClient $customers;

    /**
     * @var PlansClient $plans
     */
    public PlansClient $plans;

    /**
     * @var SubscriptionsClient $subscriptions
     */
    public SubscriptionsClient $subscriptions;

    /**
     * @var InvoicesClient $invoices
     */
    public InvoicesClient $invoices;

    /**
     * @var PaymentsClient $payments
     */
    public PaymentsClient $payments;

    /**
     * @var PaymentProvidersClient $paymentProviders
     */
    public PaymentProvidersClient $paymentProviders;

    /**
     * @var WebhooksClient $webhooks
     */
    public WebhooksClient $webhooks;

    /**
     * @var AnalyticsClient $analytics
     */
    public AnalyticsClient $analytics;

    /**
     * @var CouponsClient $coupons
     */
    public CouponsClient $coupons;

    /**
     * @var AddOnsClient $addOns
     */
    public AddOnsClient $addOns;

    /**
     * @var CreditNotesClient $creditNotes
     */
    public CreditNotesClient $creditNotes;

    /**
     * @var PortalClient $portal
     */
    public PortalClient $portal;

    /**
     * @var BillableMetricsClient $billableMetrics
     */
    public BillableMetricsClient $billableMetrics;

    /**
     * @var EventsClient $events
     */
    public EventsClient $events;

    /**
     * @var ChargesClient $charges
     */
    public ChargesClient $charges;

    /**
     * @var WalletsClient $wallets
     */
    public WalletsClient $wallets;

    /**
     * @var PaymentMethodsClient $paymentMethods
     */
    public PaymentMethodsClient $paymentMethods;

    /**
     * @var TaxesClient $taxes
     */
    public TaxesClient $taxes;

    /**
     * @var PlanOverridesClient $planOverrides
     */
    public PlanOverridesClient $planOverrides;

    /**
     * @var array{
     *   baseUrl?: string,
     *   client?: ClientInterface,
     *   maxRetries?: int,
     *   timeout?: float,
     *   headers?: array<string, string>,
     * } $options @phpstan-ignore-next-line Property is used in endpoint methods via HttpEndpointGenerator
     */
    private array $options;

    /**
     * @var RawClient $client
     */
    private RawClient $client;

    /**
     * @param string $authorization
     * @param ?string $token The token to use for authentication.
     * @param ?array{
     *   baseUrl?: string,
     *   client?: ClientInterface,
     *   maxRetries?: int,
     *   timeout?: float,
     *   headers?: array<string, string>,
     * } $options
     */
    public function __construct(
        string $authorization,
        ?string $token = null,
        ?array $options = null,
    ) {
        $defaultHeaders = [
            'Authorization' => $authorization,
            'X-Fern-Language' => 'PHP',
            'X-Fern-SDK-Name' => 'NovaBilling',
        ];
        if ($token != null) {
            $defaultHeaders['Authorization'] = "Bearer $token";
        }

        $this->options = $options ?? [];

        $this->options['headers'] = array_merge(
            $defaultHeaders,
            $this->options['headers'] ?? [],
        );

        $this->client = new RawClient(
            options: $this->options,
        );

        $this->auth = new AuthClient($this->client, $this->options);
        $this->tenants = new TenantsClient($this->client, $this->options);
        $this->apiKeys = new ApiKeysClient($this->client, $this->options);
        $this->currencies = new CurrenciesClient($this->client, $this->options);
        $this->customers = new CustomersClient($this->client, $this->options);
        $this->plans = new PlansClient($this->client, $this->options);
        $this->subscriptions = new SubscriptionsClient($this->client, $this->options);
        $this->invoices = new InvoicesClient($this->client, $this->options);
        $this->payments = new PaymentsClient($this->client, $this->options);
        $this->paymentProviders = new PaymentProvidersClient($this->client, $this->options);
        $this->webhooks = new WebhooksClient($this->client, $this->options);
        $this->analytics = new AnalyticsClient($this->client, $this->options);
        $this->coupons = new CouponsClient($this->client, $this->options);
        $this->addOns = new AddOnsClient($this->client, $this->options);
        $this->creditNotes = new CreditNotesClient($this->client, $this->options);
        $this->portal = new PortalClient($this->client, $this->options);
        $this->billableMetrics = new BillableMetricsClient($this->client, $this->options);
        $this->events = new EventsClient($this->client, $this->options);
        $this->charges = new ChargesClient($this->client, $this->options);
        $this->wallets = new WalletsClient($this->client, $this->options);
        $this->paymentMethods = new PaymentMethodsClient($this->client, $this->options);
        $this->taxes = new TaxesClient($this->client, $this->options);
        $this->planOverrides = new PlanOverridesClient($this->client, $this->options);
    }
}

File: sdks/swift/Sources/NovaBillingClient.swift
import Foundation

/// Use this class to access the different functions within the SDK. You can instantiate any number of clients with different configuration that will propagate to these functions.
public final class NovaBillingClient: Sendable {
    public let auth: AuthClient
    public let tenants: TenantsClient
    public let apiKeys: ApiKeysClient
    public let currencies: CurrenciesClient
    public let customers: CustomersClient
    public let plans: PlansClient
    public let subscriptions: SubscriptionsClient
    public let invoices: InvoicesClient
    public let payments: PaymentsClient
    public let paymentProviders: PaymentProvidersClient
    public let webhooks: WebhooksClient
    public let analytics: AnalyticsClient
    public let coupons: CouponsClient
    public let addOns: AddOnsClient
    public let creditNotes: CreditNotesClient
    public let portal: PortalClient
    public let billableMetrics: BillableMetricsClient
    public let events: EventsClient
    public let charges: ChargesClient
    public let wallets: WalletsClient
    public let paymentMethods: PaymentMethodsClient
    public let taxes: TaxesClient
    public let planOverrides: PlanOverridesClient
    private let httpClient: HTTPClient

    /// Initialize the client with the specified configuration and a static bearer token.
    ///
    /// - Parameter baseURL: The base URL to use for requests from the client. If not provided, the default base URL will be used.
    /// - Parameter token: Bearer token for authentication. If provided, will be sent as "Bearer {token}" in Authorization header.
    /// - Parameter headers: Additional headers to send with each request.
    /// - Parameter timeout: Request timeout in seconds. Defaults to 60 seconds. Ignored if a custom `urlSession` is provided.
    /// - Parameter maxRetries: Maximum number of retries for failed requests. Defaults to 2.
    /// - Parameter urlSession: Custom `URLSession` to use for requests. If not provided, a default session will be created with the specified timeout.
    public convenience init(
        baseURL: String = NovaBillingEnvironment.default.rawValue,
        token: String? = nil,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        self.init(
            baseURL: baseURL,
            headerAuth: nil,
            bearerAuth: token.map {
                .init(token: .staticToken($0))
            },
            basicAuth: nil,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
    }

    /// Initialize the client with the specified configuration and an async bearer token provider.
    ///
    /// - Parameter baseURL: The base URL to use for requests from the client. If not provided, the default base URL will be used.
    /// - Parameter token: An async function that returns the bearer token for authentication. If provided, will be sent as "Bearer {token}" in Authorization header.
    /// - Parameter headers: Additional headers to send with each request.
    /// - Parameter timeout: Request timeout in seconds. Defaults to 60 seconds. Ignored if a custom `urlSession` is provided.
    /// - Parameter maxRetries: Maximum number of retries for failed requests. Defaults to 2.
    /// - Parameter urlSession: Custom `URLSession` to use for requests. If not provided, a default session will be created with the specified timeout.
    public convenience init(
        baseURL: String = NovaBillingEnvironment.default.rawValue,
        token: ClientConfig.CredentialProvider? = nil,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        self.init(
            baseURL: baseURL,
            headerAuth: nil,
            bearerAuth: token.map {
                .init(token: .provider($0))
            },
            basicAuth: nil,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
    }

    init(
        baseURL: String,
        headerAuth: ClientConfig.HeaderAuth? = nil,
        bearerAuth: ClientConfig.BearerAuth? = nil,
        basicAuth: ClientConfig.BasicAuth? = nil,
        headers: [String: String]? = nil,
        timeout: Int? = nil,
        maxRetries: Int? = nil,
        urlSession: Networking.URLSession? = nil
    ) {
        let config = ClientConfig(
            baseURL: baseURL,
            headerAuth: headerAuth,
            bearerAuth: bearerAuth,
            basicAuth: basicAuth,
            headers: headers,
            timeout: timeout,
            maxRetries: maxRetries,
            urlSession: urlSession
        )
        self.auth = AuthClient(config: config)
        self.tenants = TenantsClient(config: config)
        self.apiKeys = ApiKeysClient(config: config)
        self.currencies = CurrenciesClient(config: config)
        self.customers = CustomersClient(config: config)
        self.plans = PlansClient(config: config)
        self.subscriptions = SubscriptionsClient(config: config)
        self.invoices = InvoicesClient(config: config)
        self.payments = PaymentsClient(config: config)
        self.paymentProviders = PaymentProvidersClient(config: config)
        self.webhooks = WebhooksClient(config: config)
        self.analytics = AnalyticsClient(config: config)
        self.coupons = CouponsClient(config: config)
        self.addOns = AddOnsClient(config: config)
        self.creditNotes = CreditNotesClient(config: config)
        self.portal = PortalClient(config: config)
        self.billableMetrics = BillableMetricsClient(config: config)
        self.events = EventsClient(config: config)
        self.charges = ChargesClient(config: config)
        self.wallets = WalletsClient(config: config)
        self.paymentMethods = PaymentMethodsClient(config: config)
        self.taxes = TaxesClient(config: config)
        self.planOverrides = PlanOverridesClient(config: config)
        self.httpClient = HTTPClient(config: config)
    }
}